File: common/src/main/java/org/octopusden/octopus/jira/api/exception/JiraApiException.java
package org.octopusden.octopus.jira.api.exception;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

public abstract class JiraApiException extends RuntimeException {

    public static final Map<String, Function<String, JiraApiException>> CODE_EXCEPTION_MAP;

    static {
        Map<String, Function<String, JiraApiException>> map = new HashMap<>();
        map.put("API-40000", BadRequestException::new);
        map.put("API-40001", FailedGenerateIPSException::new);
        CODE_EXCEPTION_MAP = Collections.unmodifiableMap(map);
    }

    private final String code;

    protected JiraApiException(String message, String code) {
        super(message);
        this.code = code;
    }

    public String getCode() {
        return code;
    }
}

File: common/src/main/kotlin/org/octopusden/octopus/jira/api/exception/JiraApiException.kt
package org.octopusden.octopus.jira.api.exception

abstract class JiraApiException(message: String, val code: String) : RuntimeException(message) {
    companion object {
        val CODE_EXCEPTION_MAP = mapOf(
            "API-40000" to { message: String -> BadRequestException(message) },
            "API-40001" to { message: String -> FailedGenerateIPSException(message) },
        )
    }
}

class BadRequestException(message: String) : JiraApiException(message, "API-40000")
class FailedGenerateIPSException(message: String) : JiraApiException(message, "API-40001")
